### PR TITLE
Use secrets when available for expiration events

### PR DESCRIPTION
**Pull request by ghislainbourgeois**
# Description

`TLScertificatesRequiresV2` will now use secrets when available instead of `update_status` to generate expiration events.

## Checklist

- [x] My code follows the [style guidelines](/CONTRIBUTING.md) of this project.
- [x] I have performed a self-review of my own code.
- [ ] I have made corresponding changes to the documentation.
- [x] I have added tests that validate the behaviour of the software.
- [x] I validated that new and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [x] I have bumped the version of any required library.

### CODE REVIEW

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py [hunk 1]
```diff
@@ -1174,10 +1181,18 @@ def _provider_certificates(self) -> List[Dict[str, str]]:
...
+            logger.debug("No relation: %s", self.relationship_name)
+            return []
         if not relation.app:
-            raise RuntimeError(f"Remote app for relation {self.relationship_name} does not exist")
+            logger.debug("No remote app in relation: %s", self.relationship_name)
+            return []
         provider_relation_data = _load_relation_data(relation.data[relation.app])
+        if not self._relation_data_is_valid(provider_relation_data):
+            logger.warning(
+                "Provider relation data did not pass JSON Schema validation: %s",
```
Comment (gruyaume) [inline lib/charms/tls_certificates_interface/v2/tls_certificates.py:1192]: We probably don't need to specify the content of the relation data in the logs.

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py [hunk 2]
```diff
@@ -1341,13 +1348,39 @@ def _on_relation_changed(self, event: RelationChangedEvent) -> None:
                         chain=certificate["chain"],
                     )
                 else:
+                    if JujuVersion.from_environ().has_secrets:
```
Comment (gruyaume) [inline lib/charms/tls_certificates_interface/v2/tls_certificates.py:1351]: `_on_relation_changed` is getting quite complex, I believe extending the docstring to explain what we're doing may be useful.

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py [hunk 3]
```diff
@@ -1341,13 +1348,39 @@ def _on_relation_changed(self, event: RelationChangedEvent) -> None:
...
+                                ),
+                            )
                     self.on.certificate_available.emit(
                         certificate_signing_request=certificate["certificate_signing_request"],
                         certificate=certificate["certificate"],
                         ca=certificate["ca"],
                         chain=certificate["chain"],
                     )
 
+    def _get_next_secret_expiry_time(self, certificate: str) -> Optional[datetime]:
```
Comment (gruyaume) [inline lib/charms/tls_certificates_interface/v2/tls_certificates.py:1377]: A docstring would be useful here too

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py [hunk 4]
```diff
@@ -1360,12 +1393,70 @@ def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
...
+        if not expiry_time:
+            # A secret expired but matching certificate is invalid. Cleaning up
+            event.secret.remove_all_revisions()
             return
-        self.on.all_certificates_invalidated.emit()
+
+        secret_expiry = event.secret.get_info().expires
+        # Satisfy type checking by ensuring the secret as an expiry time set.
+        # This should always be the case, as the secret just expired.
+        assert secret_expiry is not None
```
Comment (gruyaume) [inline lib/charms/tls_certificates_interface/v2/tls_certificates.py:1432]: Why this assert?

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py [hunk 5]
```diff
@@ -1073,7 +1077,7 @@ def _on_relation_changed(self, event: RelationChangedEvent) -> None:
         provider_relation_data = _load_relation_data(event.relation.data[self.charm.app])
         if not self._relation_data_is_valid(requirer_relation_data):
             logger.warning(
-                f"Relation data did not pass JSON Schema validation: {requirer_relation_data}"
+                "Relation data did not pass JSON Schema validation: %s", requirer_relation_data
```
Comment (gruyaume) [inline lib/charms/tls_certificates_interface/v2/tls_certificates.py:1080]: We probably don't need to specify the content of the relation data in the logs.

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py (hunk 4 above)
Comment (ghislainbourgeois) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py:1432]: The comment above explains it. It is only for type checking.

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py (hunk 5 above)
Comment (ghislainbourgeois) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py]: Agreed, removed.

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py (hunk 3 above)
Comment (ghislainbourgeois) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py:1381]: Added, please take a look at it.

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py (hunk 2 above)
Comment (ghislainbourgeois) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py:1355]: Added, please take a look at it.

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py (hunk 1 above)
Comment (ghislainbourgeois) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py]: Agreed, removed

**Review by gruyaume — APPROVED** [review]

**Pull request merged**